File: src/java/magma/app/rule/divide/DivideRule.java
package magma.app.rule.divide;

import magma.CompileError;
import magma.api.Ok;
import magma.api.Result;
import magma.app.node.MapNode;
import magma.app.node.Node;
import magma.app.rule.Rule;

import java.util.ArrayList;
import java.util.List;

public record DivideRule(String key, Rule rule) implements Rule {
    private static List<String> divide(CharSequence input) {
        DivideState current = new MutableDivideState();
        for (var i = 0; i < input.length(); i++) {
            final var c = input.charAt(i);
            current = fold(current, c);
        }

        return current.advance()
                .segments();
    }

    private static DivideState fold(DivideState divideState, char c) {
        final var appended = divideState.append(c);
        if (c == ';')
            return appended.advance();

        return appended;
    }

    @Override
    public Result<Node, CompileError> lex(String input) {
        return divide(input).stream()
                .reduce(new Ok<>(new ArrayList<>()), this::foldElement, (_, next) -> next)
                .mapValue(children -> new MapNode().withNodeList(this.key(), children));
    }

    private Result<List<Node>, CompileError> foldElement(Result<List<Node>, CompileError> maybeCurrent, String element) {
        return maybeCurrent.flatMap(current -> DivideRule.this.rule.lex(element)
                .mapValue(result -> {
                    current.add(result);
                    return current;
                }));
    }

    @Override
    public Result<String, CompileError> generate(Node node) {
        return node.findNodeList(this.key)
                .orElse(new ArrayList<>())
                .stream()
                .reduce(new Ok<>(new StringBuilder()), this::foldString, (_, next) -> next)
                .mapValue(StringBuilder::toString);
    }

    private Result<StringBuilder, CompileError> foldString(Result<StringBuilder, CompileError> maybeCurrent, Node element) {
        return maybeCurrent.flatMap(current -> this.rule.generate(element)
                .mapValue(current::append));
    }
}

File: src/node/magma/rule/divide/DivideRule.ts
/*import magma.error.CompileError;*/
/*import magma.node.MapNode;*/
/*import magma.node.Node;*/
/*import magma.result.Err;*/
/*import magma.result.Ok;*/
/*import magma.result.Result;*/
/*import magma.rule.Rule;*/
/*import java.util.Collections;*/
/*import java.util.Optional;*/
/*import java.util.stream.Collectors;*/
/*public record DivideRule(String key, Rule rule) implements Rule {
	private static DivideState divide(final CharSequence input) {
		final var length = input.length();
		DivideState current = new MutableDivideState();
		for (var i = 0; i < length; i++) {
			final var c = input.charAt(i);
			current = DivideRule.fold(current, c);
		}

		return current.advance();
	}

	private static DivideState fold(final DivideState state, final char c) {
		final var appended = state.append(c);
		if (';' == c && appended.isLevel()) return appended.advance();
		if ('{' == c) return appended.enter();
		if ('}' == c) return appended.exit();
		return appended;
	}

	private Optional<Node> lex0(final String input) {
		final var children = DivideRule.divide(input).stream().map(input1 -> this.rule.lex(input1).findValue()).flatMap(Optional::stream).toList();

		return Optional.of(new MapNode().withNodeList(this.key, children));
	}

	private Optional<String> generate0(final Node node) {
		return Optional.of(node.findNodeList(this.key)
													 .orElse(Collections.emptyList())
													 .stream()
													 .map(node1 -> this.rule().generate(node1).findValue())
													 .flatMap(Optional::stream)
													 .collect(Collectors.joining()));
	}

	@Override
	public Result<Node, CompileError> lex(final String input) {
		return this.lex0(input).<Result<Node, CompileError>>map(Ok::new).orElseGet(() -> new Err<>(new CompileError()));
	}

	@Override
	public Result<String, CompileError> generate(final Node node) {
		return this.generate0(node).<Result<String, CompileError>>map(Ok::new).orElseGet(() -> new Err<>(new CompileError()));
	}
}*/
